pkg/util/validation/limits.go

```diff
@@ -267,28 +270,35 @@ func (l *Limits) RegisterFlags(f *flag.FlagSet) {
 
 // UnmarshalYAML implements the yaml.Unmarshaler interface.
 func (l *Limits) UnmarshalYAML(value *yaml.Node) error {
-	// We want to set l to the defaults and then overwrite it with the input.
-	// To make unmarshal fill the plain data struct rather than calling UnmarshalYAML
-	// again, we have to hide it using a type indirection.  See prometheus/config.
-
-	// During startup we wont have a default value so we don't want to overwrite them
-	if defaultLimits != nil {
-		*l = *defaultLimits
-		// Make copy of default limits. Otherwise unmarshalling would modify map in default limits.
-		l.copyNotificationIntegrationLimits(defaultLimits.NotificationRateLimitPerIntegration)
-	}
-	type plain Limits
-
-	err := value.DecodeWithOptions((*plain)(l), yaml.DecodeOptions{KnownFields: true})
-	if err != nil {
-		return err
-	}
-
-	return l.validate()
+	return l.unmarshal(func(typ any) error {
+		return value.DecodeWithOptions(typ, yaml.DecodeOptions{KnownFields: true})
+	})
 }
 
 // UnmarshalJSON implements the json.Unmarshaler interface.
 func (l *Limits) UnmarshalJSON(data []byte) error {
+	return l.unmarshal(func(typ any) error {
+		dec := json.NewDecoder(bytes.NewReader(data))
+		dec.DisallowUnknownFields()
+
+		return dec.Decode(typ)
+	})
+}
+
+// unmarshal does both YAML and JSON.
+func (l *Limits) unmarshal(decode func(any) error) error {
+	// First, decode the extensions ignoring the rest of the config (if any registered).
+	var extensions map[string]interface{}
+	if len(registeredExtensions) > 0 {
+		extensionsOnlyConfig, getExtensions := newExtensionsOnlyLimitsConfig()
+
+		err := decode(extensionsOnlyConfig)
+		if err != nil {
+			return fmt.Errorf("can't decode extensions: %w", err)
+		}
+		extensions = getExtensions()
+	}
+
 	// Like the YAML method above, we want to set l to the defaults and then overwrite
```

Comment:
This comment "Like the YAML method above" no longer makes sense.

Reply:
Yup, updated in https://github.com/grafana/mimir/pull/4397/commits/34eac3f4b2840a1112d2b9ca2b2b8cc49926b161